torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -967,12 +967,37 @@ void InitXlaModuleBindings(py::module m) {
         [](const at::Tensor& tensor) { return GetTensorViewAliasId(tensor); });
   m.def("_xla_get_tensor_id",
         [](const at::Tensor& tensor) { return GetTensorId(tensor); });
-  m.def("_xla_get_devices",
-        []() { return xla::GetComputationClient()->GetLocalDevices(); });
-  m.def("_xla_num_devices",
-        []() { return xla::GetComputationClient()->GetNumDevices(); });
-  m.def("_xla_get_all_devices",
-        []() { return xla::GetComputationClient()->GetAllDevices(); });
+  m.def("_xla_get_devices", []() {
+    if (ShardingUtil::UseVirtualDevice()) {
+      // Under SPMD context, there is only one virtual devices from user
+      // perspective.
+      std::vector<std::string> all_devices =
+          xla::GetComputationClient()->GetAllDevices();
+      all_devices.resize(1);
+      return all_devices;
+    } else {
+      return xla::GetComputationClient()->GetLocalDevices();
+    }
+  });
+  m.def("_xla_num_devices", []() -> int64_t {
+    if (ShardingUtil::UseVirtualDevice()) {
+      return 1;
+    } else {
+      return xla::GetComputationClient()->GetNumDevices();
+    }
+  });
+  m.def("_xla_get_all_devices", []() {
```

Comment:
In the other thread you mentioned: `num_devices -> number of xla device` and `another concept is num_real_devices which will return the actual TPU devices`. Just want to check if my understanding is correct:

- `num_devices` refers to the logical device. So in a v4-8 non-spmd case, the number should be 4.
- `num_real_devices` refers to the physical device. In a v4-8 non-spmd case, the number should be 8?

Reply:
`num_devices` should return 1 and `num_real_devices` should reutrn 4. Since there is 1 logic(virtual) device, and 4 real device. We don't need to worry about core vs chip.